client/finality-grandpa/src/authorities.rs

```diff
@@ -631,6 +689,45 @@ impl<H, N: Add<Output=N> + Clone> PendingChange<H, N> {
 	}
 }
 
+// Tracks historical authority set changes. We store the block numbers for the first block of each
+// authority set, once they have been finalalized.
+#[derive(Debug, Encode, Decode, Clone, PartialEq)]
+pub struct AuthoritySetChanges<N>(pub Vec<(u64, N)>);
+
+impl<N: Ord + Clone> AuthoritySetChanges<N> {
+	pub(crate) fn empty() -> Self {
+		Self(Default::default())
+	}
+
+	pub(crate) fn append(&mut self, set_id: u64, block_number: N) {
+		self.0.push((set_id, block_number));
+	}
+
+	pub(crate) fn get_set_id(&self, block_number: N) -> Option<(u64, N)> {
+		let idx = self.0
+			.binary_search_by_key(&block_number, |(_, n)| n.clone())
+			.unwrap_or_else(|b| b);
+		if idx < self.0.len() {
+			let (set_id, block_number) = self.0[idx].clone();
+			// To make sure we have the right set we need to check that the one before it also exists.
+			if idx > 0 {
+				let (prev_set_id, _) = self.0[idx - 1usize];
+				if set_id != prev_set_id + 1u64 {
```

Comment:
How can that fail? Wouldn't that mean that the entire object is wrong?

Reply:
Yeah pretty much, if this happens the object is missing authority set changes and something must have gone wrong applying changes. It's my understanding that this can't happen, but I'm not certain enough of that to omit this check ... @andresilva what do you think?

Reply:
We wanted to fail gracefully for nodes that will migrate to this and therefore won't have all set ids stored here (only when they sync from scratch), the only check that's really needed for this is the first one `idx < self.0.len()`. These checks don't hurt but they can be removed since they would imply that this data structure skipped set ids spuriously which is wrong.